Mathlib/CategoryTheory/Limits/Shapes/Types.lean

```diff
@@ -720,12 +720,202 @@ end Pullback
+
+/-- Construction of the pushout in the category of types, as a quotient of `X₁ ⊕ X₂`. -/
+def Pushout : Type u := _root_.Quot (Pushout.Rel f g)
+
+/-- In case `f : S ⟶ X₁` is a monomorphism, this relation is the equivalence relation
+generated by `Pushout.Rel f g`. -/
+inductive Pushout.Rel' : X₁ ⊕ X₂ → X₁ ⊕ X₂ → Prop
+  | refl (x : X₁ ⊕ X₂) : Rel' x x
+  | inl_inl (x₀ y₀ : S) (h : g x₀ = g y₀) : Rel' (Sum.inl (f x₀)) (Sum.inl (f y₀))
+  | inl_inr (s : S) : Rel' (Sum.inl (f s)) (Sum.inr (g s))
+  | inr_inl (s : S) : Rel' (Sum.inr (g s)) (Sum.inl (f s))
+
+/-- The quotient of `X₁ ⊕ X₂` by the relation `PushoutRel' f g`. -/
+def Pushout' : Type u := _root_.Quot (Pushout.Rel' f g)
+
+namespace Pushout
+
+/-- The left inclusion in the constructed pushout `Pushout f g`. -/
+@[simp]
+def inl : X₁ ⟶ Pushout f g := fun x => Quot.mk _ (Sum.inl x)
+
+/-- The right inclusion in the constructed pushout `Pushout f g`. -/
+@[simp]
+def inr : X₂ ⟶ Pushout f g := fun x => Quot.mk _ (Sum.inr x)
+
+lemma condition : f ≫ inl f g = g ≫ inr f g := by
+  ext x
+  exact Quot.sound (Rel.inl_inr x)
+
+/-- The constructed pushout cocone in the category of types. -/
+@[simps!]
+def cocone : PushoutCocone f g := PushoutCocone.mk _ _ (condition f g)
+
+/-- The cocone `cocone f g` is colimit. -/
+def isColimitCocone : IsColimit (cocone f g) :=
+  PushoutCocone.IsColimit.mk _ (fun s => Quot.lift (fun x => match x with
+      | Sum.inl x₁ => s.inl x₁
+      | Sum.inr x₂ => s.inr x₂) (by
+    rintro _ _ ⟨t⟩
+    exact congr_fun s.condition t)) (fun _ => rfl) (fun _ => rfl) (fun s m h₁ h₂ => by
+      ext ⟨x₁|x₂⟩
+      · exact congr_fun h₁ x₁
+      · exact congr_fun h₂ x₂)
+
+@[simp]
+lemma inl_rel'_inl_iff (x₁ y₁ : X₁) :
+    Rel' f g (Sum.inl x₁) (Sum.inl y₁) ↔ x₁ = y₁ ∨
+      ∃ (x₀ y₀ : S) (_ : g x₀ = g y₀), x₁ = f x₀ ∧ y₁ = f y₀ := by
+  constructor
+  · rintro (_|⟨_, _, h⟩)
+    · exact Or.inl rfl
+    · exact Or.inr ⟨_, _, h, rfl, rfl⟩
+  · rintro (rfl | ⟨_,_ , h, rfl, rfl⟩)
+    · apply Rel'.refl
+    · exact Rel'.inl_inl _ _ h
+
+@[simp]
+lemma inl_rel'_inr_iff (x₁ : X₁) (x₂ : X₂) :
+    Rel' f g (Sum.inl x₁) (Sum.inr x₂) ↔
+      ∃ (s : S), x₁ = f s ∧ x₂ = g s := by
+  constructor
+  · rintro ⟨_⟩
+    exact ⟨_, rfl, rfl⟩
+  · rintro ⟨s, rfl, rfl⟩
+    exact Rel'.inl_inr _
+
+@[simp]
+lemma inr_rel'_inr_iff (x₂ y₂ : X₂) :
+    Rel' f g (Sum.inr x₂) (Sum.inr y₂) ↔ x₂ = y₂ := by
+  constructor
+  · rintro ⟨_⟩
+    rfl
+  · rintro rfl
+    apply Rel'.refl
+
+variable {f g}
+
+lemma Rel'.symm {x y : X₁ ⊕ X₂} (h : Rel' f g x y) :
+    Rel' f g y x := by
+  obtain _|⟨_, _, h⟩|_|_ := h
+  · apply Rel'.refl
+  · exact Rel'.inl_inl _ _ h.symm
+  · exact Rel'.inr_inl _
+  · exact Rel'.inl_inr _
+
+variable (f g)
+
+lemma equivalence_rel' [Mono f] : _root_.Equivalence (Rel' f g) where
+  refl := Rel'.refl
+  symm h := h.symm
+  trans := by
+    intro x y z hxy hyz
+    obtain _|⟨_, _, h⟩|s|_ := hxy
+    · exact hyz
+    · obtain z₁|z₂ := z
```

Comment:
Can this step also be integrated into the `rintro`?

Reply:
Thanks for the suggestions! I have use `rintro` for the `hxy` variable, but not for `z` because this would create 8 goals instead of 7.